src/screencast.rs

```diff
@@ -0,0 +1,146 @@
+use std::collections::HashMap;
+
+use zbus::dbus_interface;
+
+use zbus::zvariant::{DeserializeDict, ObjectPath, OwnedValue, SerializeDict, Type, Value};
+
+use enumflags2::BitFlags;
+
+use serde::{Deserialize, Serialize};
+
+use crate::request::RequestInterface;
+use crate::session::{append_session, CursorMode, PersistMode, Session, SourceType, SESSIONS};
+
+#[derive(SerializeDict, DeserializeDict, Type, Debug, Default)]
+/// Specified options for a [`Screencast::create_session`] request.
+#[zvariant(signature = "dict")]
+struct SessionCreateResult {
+    handle_token: String,
+}
+
+#[derive(SerializeDict, DeserializeDict, Type, Debug, Default)]
+/// Specified options for a [`Screencast::select_sources`] request.
+#[zvariant(signature = "dict")]
+pub struct SelectSourcesOptions {
+    /// A string that will be used as the last element of the handle.
+    /// What types of content to record.
+    pub types: Option<BitFlags<SourceType>>,
+    /// Whether to allow selecting multiple sources.
+    pub multiple: Option<bool>,
+    /// Determines how the cursor will be drawn in the screen cast stream.
+    pub cursor_mode: Option<CursorMode>,
+    pub restore_token: Option<String>,
+    pub persist_mode: Option<PersistMode>,
+}
+
+#[derive(Clone, Serialize, Deserialize, Type, Default, Debug)]
+/// A PipeWire stream.
+pub struct Stream(u32, StreamProperties);
+
+#[derive(Clone, SerializeDict, DeserializeDict, Default, Type, Debug)]
+/// The stream properties.
+#[zvariant(signature = "dict")]
+struct StreamProperties {
+    id: Option<String>,
+    position: Option<(i32, i32)>,
+    size: Option<(i32, i32)>,
+    source_type: Option<SourceType>,
+}
+
+// TODO: this is copy from ashpd, but the dict is a little different from xdg_desktop_portal
+#[derive(Clone, SerializeDict, DeserializeDict, Default, Debug, Type)]
+#[zvariant(signature = "dict")]
+struct StartReturnValue {
+    streams: Vec<Stream>,
+    persist_mode: u32,
+    restore_token: Option<String>,
+}
+
+pub struct ScreenCast;
+
+#[dbus_interface(name = "org.freedesktop.impl.portal.ScreenCast")]
+impl ScreenCast {
+    #[dbus_interface(property, name = "version")]
+    fn version(&self) -> u32 {
+        4
+    }
+
+    #[dbus_interface(property)]
+    fn available_cursor_modes(&self) -> u32 {
+        (CursorMode::Hidden | CursorMode::Embedded).bits()
+    }
+
+    #[dbus_interface(property)]
+    fn available_source_types(&self) -> u32 {
+        BitFlags::from_flag(SourceType::Monitor).bits()
```

Comment:
I want to support region cast later.., so  virtual maybe later? but what about window?

Reply:
Window cast will need top-level data from compositor. We should look into making our own Wayland protocol to expose these details in that case.